File: src/client/NodeJS/ms-rest-azure/lib/pollingState.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information. 
'use strict';

var util = require('util');
var msRest = require('ms-rest');
var LroStates = require('./constants').LongRunningOperationStates;

/**
 * @class
 * Initializes a new instance of the PollingState class.
 * @constructor
 * @param {object} [response] - The response object to extract long
 * running operation status.
 * 
 * @param {number} retryTimeout - The timeout in seconds to retry on
 * intermediate operation results. Default Value is 30.
 */
function PollingState(resultOfInitialRequest, retryTimeout) {
  if (retryTimeout === null || retryTimeout === undefined) {
    retryTimeout = 30;
  }
  this._retryTimeout = retryTimeout;
  this.updateResponse(resultOfInitialRequest.response);
  this.request = resultOfInitialRequest.request;
  //Parse response.body & assign it as the resource
  try {
    if (resultOfInitialRequest.body && 
        typeof resultOfInitialRequest.body.valueOf() === 'string' &&
        resultOfInitialRequest.body.length > 0) {
      this.resource = JSON.parse(resultOfInitialRequest.body);
    } else {
      this.resource = resultOfInitialRequest.body;
    } 
  } catch (error) {
    var deserializationError = new Error(util.format('Error "%s" occurred in parsing the responseBody ' + 
      'while creating the PollingState for Long Running Operation- "%s"', error, resultOfInitialRequest.body));
    deserializationError.request = resultOfInitialRequest.request;
    deserializationError.response = resultOfInitialRequest.response;
    throw deserializationError;
  }
  
  switch (this.response.statusCode) {
    case 202:
      this.status = LroStates.InProgress;
      break;

    case 204:
      this.status = LroStates.Succeeded;
      break;
    case 201:
      if (this.resource && this.resource.properties && this.resource.properties.provisioningState) {
        this.status = this.resource.properties.provisioningState;
      } else {
        this.status = LroStates.InProgress;
      }
      break;
    case 200:
      if (this.resource && this.resource.properties && this.resource.properties.provisioningState) {
        this.status = this.resource.properties.provisioningState;
      } else {
        this.status = LroStates.Succeeded;
      }
      break;
    default:
      this.status = LroStates.Failed;
      break;
  }
}

/**
 * Gets timeout in milliseconds. 
 * @returns {number} timeout
 */
PollingState.prototype.getTimeout = function () {
  if (this._retryTimeout || this._retryTimeout === 0) {
    return this._retryTimeout * 1000;
  }
  if (this.response && this.response.headers['retry-after']) {
    return parseInt(this.response.headers['retry-after']) * 1000;
  }
  return 30 * 1000;
};

/**
 * Update cached data using the provided response object
 * @param {object} [response] - provider response object.
 */
PollingState.prototype.updateResponse = function (response) {
  this.response = response;
  if (response && response.headers) {
    if (response.headers['azure-asyncoperation']) {
      this.azureAsyncOperationHeaderLink = response.headers['azure-asyncoperation'];
    }
    
    if (response.headers['location']) {
      this.locationHeaderLink = response.headers['location'];
    }
  }
};

/**
 * Returns long running operation result.
 * @returns {object} HttpOperationResponse
 */
PollingState.prototype.getOperationResponse = function () {
  var result = new msRest.HttpOperationResponse();
  result.request = this.request;
  result.response = this.response;
  if (this.resource && typeof this.resource.valueOf() === 'string') {
    result.body = this.resource;
  } else {
    result.body = JSON.stringify(this.resource);
  }
  return result;
};

/**
 * Returns an Error on operation failure.
 * @returns {object} Error
 */
PollingState.prototype.getCloudError = function (err) {
  var errMsg;
  var errCode;

  var error = new Error();
  error.request = msRest.stripRequest(this.request);
  var parsedResponse = this.response.body;
  error.response = msRest.stripResponse(this.response);
  try {
    if (this.response.body && typeof this.response.body.valueOf() === 'string') {
      if (this.response.body === '') this.response.body = null;
      parsedResponse = JSON.parse(this.response.body);
    }
  } catch (err) {
    error.message = util.format('Error "%s" occurred while deserializing the error ' + 
      'message "%s" for long running operation.', err.message, this.response.body);
    return error;
  }
  
  if (err && err.message) {
    errMsg = util.format('Long running operation failed with error: \'%s\'.', err.message);
  } else {
    errMsg = util.format('Long running operation failed with status: \'%s\'.', this.status);
  }

  if (parsedResponse) {
    if (parsedResponse.error && parsedResponse.error.message) {
      errMsg = util.format('Long running operation failed with error: \'%s\'.', parsedResponse.error.message);
    }
    if (parsedResponse.error && parsedResponse.error.code) {
      errCode = parsedResponse.error.code;
    }
  }

  error.message = errMsg;
  if (errCode) error.code = errCode;
  error.body = parsedResponse;
  return error;
};

module.exports = PollingState;

File: src/client/Ruby/ms-rest-azure/lib/ms_rest_azure/polling_state.rb
# encoding: utf-8
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

module MsRestAzure
  #
  # Class which represents a state of Azure long running operation.
  #
  class PollingState

    # @return [Net::HTTPRequest] the HTTP request.
  	attr_accessor :request

    # @return the resource
  	attr_accessor :resource

    # @return [Net::HTTPResponse] the HTTP response.
    attr_accessor :response

    # @return [AzureOperationError] the azure error data.
    attr_accessor :error_data

    # @return [String] the latest value captured from Azure-AsyncOperation header.
    attr_accessor :azure_async_operation_header_link

    # @return [String] the latest value captured from Location header.
    attr_accessor :location_header_link

    # @return [String] status of the long running operation.
    attr_accessor :status

    def initialize(azure_response, retry_timeout)
      @retry_timeout = retry_timeout
      @request = azure_response.request
      update_response(azure_response.response)
      @resource = azure_response.body

      # On non flattened resource, we should find provisioning_state inside 'properties'
      if (!@resource.nil? && @resource.respond_to?(:properties) && @resource.properties.respond_to?(:provisioning_state) && !@resource.properties.provisioning_state.nil?)
        @status = @resource.properties.provisioning_state
      # On flattened resource, we should find provisioning_state at the top level
      elsif !@resource.nil? && @resource.respond_to?(:provisioning_state) && !@resource.provisioning_state.nil?
        @status = @resource.provisioning_state
      else
        case @response.status
          when 202
            @status = AsyncOperationStatus::IN_PROGRESS_STATUS
          when 200, 201, 204
            @status = AsyncOperationStatus::SUCCESS_STATUS
          else
            @status = AsyncOperationStatus::FAILED_STATUS
          end
      end
    end

    #
    # Returns the amount of time in seconds for long running operation polling delay.
    #
    # @return [Integer] Amount of time in seconds for long running operation polling delay.
    def get_delay
      return @retry_timeout unless @retry_timeout.nil?

      if !response.nil? && !response.headers['Retry-After'].nil?
        return response.headers['Retry-After'].to_i
      end

      return AsyncOperationStatus::DEFAULT_DELAY
    end

    #
    # Updates the polling state from the fields of given response object.
    # @param response [Net::HTTPResponse] the HTTP response.
    def update_response(response)
      @response = response

      unless response.nil?
        @azure_async_operation_header_link = response.headers['Azure-AsyncOperation'] unless response.headers['Azure-AsyncOperation'].nil?
        @location_header_link = response.headers['Location'] unless response.headers['Location'].nil?
      end
    end

    #
    # returns the Azure's response.
    #
    # @return [MsRestAzure::AzureOperationResponse] Azure's response.
    def get_operation_response
      azure_response = AzureOperationResponse.new(@request, @response, @resource)
      azure_response
    end

    #
    # Composes and returns cloud error.
    #
    # @return [AzureOperationError] the cloud error.
    def get_operation_error
      AzureOperationError.new @request, @response, @error_data, "Long running operation failed with status #{@status}"
    end
    
    def get_request(options = {})
      link = @azure_async_operation_header_link || @location_header_link
      options[:connection] = create_connection(options[:base_uri])
      MsRest::HttpOperationRequest.new(nil, link, :get, options)
    end

    private

    # @return [Integer] retry timeout.
    attr_accessor :retry_timeout

    attr_accessor :connection

    def create_connection(base_url)
      @connection ||= Faraday.new(:url => base_url, :ssl => MsRest.ssl_options) do |faraday|
        [[MsRest::RetryPolicyMiddleware, times: 3, retry: 0.02], [:cookie_jar]].each{ |args| faraday.use(*args) }
        faraday.adapter Faraday.default_adapter
        faraday.headers = request.headers
        logging = ENV['AZURE_HTTP_LOGGING'] || request.log
        if logging
          faraday.response :logger, nil, { :bodies => logging == 'full' }
        end
      end
    end
  end

end